### DISCUSSION TITLE
Manual zoom on both axis(price and time)

### DISCUSSION THREAD

**Discussion by publine2** [highfestiva/finplot#207] [Q&A]
Is it possible to disable automatic zoom and do manual zoom by mouse like in TradingView?  Will be very appreciated for code example.

**Comment by highfestiva** [comment]
Currently you can only zoom in/out in Y by `Ctrl+ScrollWheel`. Apart from that it currently is not possible, as I can't imagine a case where it's better to _not_ center around price. Do you have such a case?

**Reply by highfestiva** [reply]
Though you could use RButton-drag to temporarily zoom in to something specific in certain circumstances, but I suppose that is not what you're asking for.

**Reply by publine2** [reply]
Ctrl+ScrollWheel is good, I can live with that, but when I move the chart back and forward the automatic zooming(or scaling) still making problems for me. Is it possible to switch off this automatic stuff? Normally, if you doing technical analysis on the chart, the least thing you would like to have is the scale change on every chart move. At least for me. By the way, all financial charting systems which I have ever used in my life, all of them have options to disable automatic zooming/scaling/centering and make that manual. So, probably, that manual stuff is actually needed especially for financial charts :) But anyway, thank you for reply, I just thought it should not be much harder than just overriding couple functions with empty ones.

**Reply by highfestiva** [reply]
Probably a matter of getting used to it during TA. Please file a bug report (I currently don't have feature requests), and I'll add it for v1.8.

**Reply by publine2** [reply]
I wrote some quickfix module around your finplot to solve all that problems. And as I thought before - it's not that really hard to fix that, just couple strings of code.
-------------------------finplotypan.py----------------------

    #!/usr/bin/env python3
    
    import sys
    import finplot as fplt
    from finplot import *
    
    class FinViewBoxNocsm(fplt.FinViewBox):
        def __init__(self, *args, **kwargs):
            self.y_pan=False # False = default behavior
            super().__init__(*args, **kwargs)
    
        def setMouseEnabled(self,x=True, y=False):
            super().setMouseEnabled(x=True, y=self.y_pan)
    
        def set_range(self, x0, y0, x1, y1): 
            blockedFuncs=['linkedViewChanged','refresh_all_y_zoom']
            if self.y_pan and sys._getframe(1).f_code.co_name=='update_y_zoom' and sys._getframe(2).f_code.co_name in blockedFuncs:
                vr = self.viewRect()
                y0 = vr.top()
                y1 = vr.bottom()
            super().set_range(x0,y0,x1,y1)
    
        def mouseLeftDrag(self, ev, axis):
            super().mouseLeftDrag(ev, None if self.y_pan else axis)
    
        def set_y_pan(self,state=True):
            self.y_pan = state
            self.setMouseEnabled(x=True, y=self.y_pan)
    
    fplt.FinViewBox=FinViewBoxNocsm
    
    #example of usage:
    #import finplotypan as fplt
    #ax0 = fplt.create_plot()
    #ax0.vb.set_y_pan() #disable automatic zoom, enable y pan
    #ax0.vb.set_y_pan(False) #revert to normal behavior

--------------------------

**Reply by highfestiva** [reply]
I suppose that'll disable x-linked panning between different views. Is that ok/desired?

**Reply by publine2** [reply]
Yes, that is what I wanted. It's not disable x-linking, i't just disable y-autocenter/autozoom... only for the chart where you do set_y_pan(), and you always can go to default behavior by set_y_pan(False). But that is just a "dirty hack" just for now. I believe you will make some kind a set_y_pan() option in normal programming way for v1.8.  I tested it with your dockable.py example - works perfect, exactly how it should be.

**Comment by highfestiva** [comment]
Please supply a case where manual Y zoom/pan is preferable, as I asked for initially.

If you have such as case and like the feature included in finplot, please add it to the existing class `FinViewBox`. And please keep the pull request minimalistic and clean, without any `sys._getframe` hacks or unnecessary overrides. Perhaps a hotkey for toggling on/off, and re-rendering when turning on again? Or perhaps dragging on the Y-axis as in TradingView would be better (default off)?

The reason I want these things from you is because, at it's core, finplot is for automated backtesting and inspection, not for manual TA. If we can make it better for manual traders with minimal adjustments, that's great! If not, well then those features are for another lib — or your own fork.
